### DISCUSSION TITLE
What is the purpose of pterm.PrintOnErrorf?

### DISCUSSION THREAD

**Discussion by sheldonhull** [pterm/pterm#301] [Q&A] [answered]
```go
func TestPrintOnErrorf(t *testing.T) {
t.Run("PrintOnErrorf", func(t *testing.T) {
result := captureStdout(func(w io.Writer) {
pterm.PrintOnErrorf("wrapping error : %w", errors.New("hello world"))
})
testza.AssertContains(t, result, "hello world")
})
}
```

I'm working on improving my usage of Pterm and would like to know what the purpose of using a function like `pterm.PrintOnErrorf` is. 

My assumption would be if I was using something like this I'd be writing to console while also returning an error so that it could be passed back up like:

```
func DoSomething() error {

err := doSomething()
if err != nil { 
return pterm.PrintOnErrorf("wrapping error : %w", errors.New("doSomething() unable to connect"))
}
return nil 
}
```

As it stands the value seems to wrap, but not return, so I'm a little confused on why I'd use this. What am I missing?

**Comment by MarvinJWendt** [comment, marked as answer]
Hi @sheldonhull, after all `PrintOnErrorf(abc, x)` is just a wrapper for `PrintOnError(fmt.Errorf(abc, x))`. The main reason where you'd use this function is in CLIs, where you want to notify the user about an error. The error should be handled after calling that function. 

Here's an example:

```go
func main() {
	pterm.Println("Hello, world!")
	err := errors.New("some error from library")
	pterm.Error.PrintOnErrorf("There was an error when starting the CLI: %w", err)
	// Handle error here
	pterm.Info.Println("This is an info message")
}
```

![image](https://user-images.githubusercontent.com/31022056/147391504-77f92466-e387-49f7-b029-df0df206bf53.png)

Here is what happens when the error is nil:
```go
func main() {
	pterm.Println("Hello, world!")
	var err error // here the error is nil
	pterm.Error.PrintOnErrorf("There was an error when starting the CLI: %w", err)
	// Handle error here
	pterm.Info.Println("This is an info message")
}
```

![image](https://user-images.githubusercontent.com/31022056/147391512-6af7a0d9-10c7-403e-b309-67300bc217d7.png)


As you can see, the message is only printed when the error is not nil. The formatting in `PrintOnErrorf` is just to make it easier to describe an error to the user and add more information to the message.

Another thing would be to use `pterm.Fatal.PrintOnErrorf(abc, x)`. This would panic the application directly (due to the Fatal PrefixPrinter) when an error is not nil and add some information to it. More or less a shortcut for:

```go
err := errors.New("test")
if err != nil {
	pterm.Fatal.Printf(abc, x)
}
```

So in general, you would use this function where you panic an error.

```go
output, err := someFunction()
if err != nil {
	panic(fmt.Errorf("There was an error while doing XXX: %w", err))
}

// -->
output, err := someFunction()
pterm.Fatal.PrintOnErrorf("There was an error while doing XXX: %w", err)

// Or if you just want to let the user know, without panicing:

output, err := someFunction()
pterm.Warning.PrintOnErrorf("There was an error while doing XXX: %w", err)
           |-- Use the appropriate TextPrinter here (Warning, Info, Debug, etc)

```

I hope this helps :)
